portability-transfer/src/main/java/org/dataportabilityproject/transfer/PortabilityInMemoryDataCopier.java

```diff
@@ -121,16 +122,72 @@ private void copyHelper(
             exportAuthData,
             importAuthData,
             Optional.of(new ExportInformation(
-                continuationData.getPaginationData(), exportInformation.get().getContainerResource())));
+                continuationData.getPaginationData(),
+                exportInformation.get().getContainerResource())));
       }
 
       // Start processing sub-resources
       if (continuationData.getContainerResources() != null
           && !continuationData.getContainerResources().isEmpty()) {
         for (ContainerResource resource : continuationData.getContainerResources()) {
-          copyHelper(jobId, exportAuthData, importAuthData, Optional.of(new ExportInformation(null, resource)));
+          copyHelper(jobId, exportAuthData, importAuthData,
+              Optional.of(new ExportInformation(null, resource)));
+        }
+      }
+    }
+  }
+
+  private ExportResult runExportLogic(UUID jobId, AuthData exportAuthData,
+      Optional<ExportInformation> exportInformation) {
+    ExportResult<?> exportResult = exportHelper(jobId, exportAuthData, exportInformation);
+    if (exportResult.getType() == ResultType.ERROR) {
+      ExceptionResponse response = checkRetry(exportResult.getMessage());
+      if (response.canRetry) {
+        int retries = 0;
+        // TODO: what if different kinds of errors have different max retries, and we start with one
+        // kind of error and then start seeing another?
+        while (retries < response.maxRetries && response.canRetry) {
+          exportResult = exportHelper(jobId, exportAuthData, exportInformation);
+          if (exportResult.getType() != ResultType.ERROR) {
+            return exportResult;
+          } else {
+            retries += 1;
+            response = checkRetry(exportResult.getMessage());
+          }
         }
+        return exportResult;
+      } else {
+        return exportResult;
       }
     }
+    return exportResult;
+  }
+
+  private ExportResult exportHelper(UUID jobId, AuthData exportAuthData,
+      Optional<ExportInformation> exportInformation) {
+    logger.debug("Starting export");
+    ExportResult<?> exportResult = exporter.get().export(jobId, exportAuthData, exportInformation);
+    logger.debug("Finishing export");
+    return exportResult;
   }
+
+  private ExceptionResponse checkRetry(String exceptionMessage) {
```

Comment:
can we also check based on http status code?

Reply:
I think that should wait until v1.